livekit-plugins/livekit-plugins-assemblyai/livekit/plugins/assemblyai/stt.py

```
@@ -0,0 +1,423 @@
+
+    async def _run_ws(self, ws: aiohttp.ClientWebSocketResponse) -> None:
+        """
+        This method can throw ws errors, these are handled inside the _run method
+        """
+
+        closing_ws = False
+        self._queue.put_nowait(self._END_UTTERANCE_SILENCE_THRESHOLD_MSG)
+
+        # Local variables for buffering
+        buffer = bytearray()
+        buffer_duration = 0.0
+
+        async def send_task():
+            nonlocal closing_ws, buffer, buffer_duration
+            # forward inputs to AssemblyAI
+            # if we receive a close message, signal it to AssemblyAI and break.
+            # the recv task will then make sure to process the remaining audio and stop
+            while True:
+                data = await self._queue.get()
+                self._queue.task_done()
+
+                if isinstance(data, rtc.AudioFrame):
+                    # TODO: The remix_and_resample method is low quality
+                    # and should be replaced with a continuous resampling
+                    frame = data.remix_and_resample(
+                        self._sample_rate,
+                        self._num_channels,
+                    )
+                    buffer.extend(frame.data.tobytes())
+
+                    # Calculate buffer duration
+                    total_frames = len(buffer) / bytes_per_frame[self._opts.encoding]
+                    samples_per_second = self._sample_rate * self._num_channels
+                    buffer_duration = total_frames / samples_per_second
+
+                    if buffer_duration >= self._buffer_size_seconds:
+                        await ws.send_bytes(bytes(buffer))
+                        buffer.clear()
+                        buffer_duration = 0.0
+                elif data == SpeechStream._CLOSE_MSG:
+                    closing_ws = True
+                    await ws.send_str(data)  # tell AssemblyAI we are done with inputs
+                    break
```

should we add `else` to raise Exception if there was an unexpected data type?